plugins/markdown/core/src/org/intellij/plugins/markdown/ui/preview/jcef/CodeFenceLanguageParsingSupport.kt

```diff
@@ -0,0 +1,279 @@
+// Copyright 2000-2025 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
+package org.intellij.plugins.markdown.ui.preview.jcef
+
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.startup.ProjectActivity
+import com.intellij.ui.jcef.JBCefBrowser
+import org.cef.CefSettings
+import org.cef.browser.CefBrowser
+import org.cef.handler.CefDisplayHandlerAdapter
+import org.intellij.lang.annotations.Language
+import org.intellij.markdown.ast.ASTNode
+import org.intellij.markdown.html.HtmlGenerator
+import org.intellij.plugins.markdown.ui.preview.html.DefaultCodeFenceGeneratingProvider
+import org.intellij.plugins.markdown.ui.preview.jcef.impl.executeJavaScript
+import org.intellij.plugins.markdown.ui.preview.jcef.impl.waitForPageLoad
+import java.net.URL
+import java.util.concurrent.CountDownLatch
+import java.util.concurrent.TimeUnit
+import kotlin.random.Random
+
+class CodeFenceLanguageParsingSupport : ProjectActivity {
+  override suspend fun execute(project: Project) {
+    codeFenceParsingStartUp()
+  }
+
+  companion object {
+    @Volatile private var started = false
+    @Volatile private var startupCompleted = false
+    @Volatile private var available = true
+    @Volatile private var startupLatch: CountDownLatch? = CountDownLatch(1)
+    @Volatile private var jsExecLatch: CountDownLatch? = null
+    @Volatile private var jsResultLatch: CountDownLatch? = null
+    @Volatile private var jsResult: String? = null
+    @Volatile private var jsError = false
+
+    private val browser = object : JBCefBrowser() {
+      init {
+        jbCefClient.addDisplayHandler(object: CefDisplayHandlerAdapter() {
+          override fun onConsoleMessage(browser: CefBrowser, level: CefSettings.LogSeverity, message: String, source: String, line: Int): Boolean {
+            if (level == CefSettings.LogSeverity.LOGSEVERITY_INFO && message.startsWith("highlighter:")) {
+              jsExecLatch?.await()
+              jsResult = message.substring(12)
+              jsResultLatch?.countDown()
+            }
+            else if (level == CefSettings.LogSeverity.LOGSEVERITY_ERROR) {
+              System.err.println("Error while executing highlighter: $message")
+              jsError = true
+            }
+
+            return true // Prevent creation of jcef_*.log files
+          }
+        }, cefBrowser)
+      }
+    }
+
+    internal suspend fun codeFenceParsingStartUp() {
+      if (started)
+        return
+
+      started = true
+
+      var latestJS: String? = null
+      var fallbackJS: String? = null
+      val scriptLatch = CountDownLatch(2)
+
+      Thread {
+        try {
+          latestJS = URL("https://unpkg.com/@highlightjs/cdn-assets/highlight.min.js").readText()
```

Comment:
I'd definitely understand if you would prefer to always use a fixed version of highlight.js instead of downloading the latest version like this.